### PR TITLE
logger unit tests

### LINKED ISSUE #5: Add unit tests
Add some unit tests to the project

##### time: < 1 hour

_original context https://github.com/ubiquity/ubiquibot-logger/pull/4#issuecomment-2180461549_

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves #5
Requires #4

### CODE REVIEW

**Comment by Keyrxng** [comment]
@gentlementlegen added some tests covering things. I've also decided to remove `util` and just go with `JSON.stringify`. 

The difference I could see writing these simple tests was not significant enough to remain using it when it may be a problem down the line. Essentially it was including things like array length and such but I don't feel this is awfully useful information when weighing the pros and cons.

**Review by 0x4007 — APPROVED** [review]
Looks fine. You can merge if @gentlementlegen doesn't get around to this soon.

File: .github/workflows/jest-testing.yml
Comment (gentlementlegen) [inline .github/workflows/jest-testing.yml:1]: Has to be updated to avoid comment spam. Refer to https://github.com/ubiquity/ts-template/blob/development/.github/workflows/jest-testing.yml

File: jest.config.ts
Comment (gentlementlegen) [inline jest.config.ts:1]: See https://github.com/ubiquity/ts-template/blob/development/jest.config.json (also need to add the related package)

File: package.json [hunk 1]
```diff
@@ -90,5 +95,6 @@
     "extends": [
       "@commitlint/config-conventional"
     ]
-  }
+  },
+  "packageManager": "yarn@1.22.22+sha1.ac34549e6aa8e7ead463a7407e1c7390f61a6610"
```
Comment (gentlementlegen) [inline package.json]: Do we need the sha here?

File: src/logs.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+      metadata = {};
+    }
+    if (typeof metadata == "string" || typeof metadata == "number") {
+      // TODO: think i need to support every data type
+      metadata = { message: metadata };
+    }
+
+    const stackLines = new Error().stack?.split("\n") || [];
+    if (stackLines.length > 3) {
+      const callerLine = stackLines[3]; // .replace(process.cwd(), "");
```
Comment (gentlementlegen) [inline src/logs.ts]: Remove comment probably

File: src/logs.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  private _addDiagnosticInformation(metadata: Record<string, unknown> | string | number | null | undefined) {
+    // this is a utility function to get the name of the function that called the log
+    // I have mixed feelings on this because it manipulates metadata later possibly without the developer understanding why and where,
+    // but seems useful for the metadata parser to understand where the comment originated from
+
+    if (!metadata) {
+      metadata = {};
+    }
+    if (typeof metadata == "string" || typeof metadata == "number") {
+      // TODO: think i need to support every data type
```
Comment (gentlementlegen) [inline src/logs.ts]: Can this TODO be solved beforehand?

File: src/logs.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+      return ["<!--", prettySerialized, "-->"].join("\n");
+    }
+  }
+
+  private _diffColorCommentMessage(type: string, message: string) {
+    const diffPrefix = {
+      fatal: "-", // - text in red
+      ok: "+", // + text in green
+      error: "!", // ! text in orange
+      // info: "#", // # text in gray
```
Comment (gentlementlegen) [inline src/logs.ts]: Why all the comments?

File: src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+// eslint-disable-next-line no-control-regex
```
Comment (gentlementlegen) [inline src/utils.ts:1]: Why is this required? Also this file doesn't seem to be properly formatted.

**Review by gentlementlegen — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Keyrxng) [reply package.json]: After updating my env if this field is not defined then it will be defined automatically by yarn during install and it appends the sha itself. So because of this I left it as is but I don't think it's strictly necessary is it, I will remove it as it's not present in repos using V4 either.

File: src/logs.ts (hunk 4 above)
Comment (Keyrxng) [reply src/logs.ts]: These are comments from the original I thought best to leave them but I will do a thorough clean up of those that are not needed

File: src/logs.ts (hunk 3 above)
Comment (Keyrxng) [reply src/logs.ts]: Well it's not required to handle any other data types really. `Metadata` has `[key: string]: unknown;` so it can already process every data type so long as it's a property of metadata which is cleaner than allowing and having to process others isn't it?

So I think just a clean up here is needed because `_addDiagnosticInformation()` should only really be passed a `Metadata` object or undefined as that's what the public log functions accept as input.

This way invoking logs should just be `.fatal("xyz error", { error: e, a: [], b: {}, ...}`

File: package.json (hunk 1 above)
Comment (gentlementlegen) [reply package.json]: I am not sure about the proper course to chose as we do not use yarn 4 extensively in our repositories. As long as it works and that it is supported I am fine with it, but I think it shall work with only `yarn@1.22.22`

File: src/utils.ts (hunk 5 above)
Comment (Keyrxng) [reply src/utils.ts:1]: https://stackoverflow.com/questions/49743842/javascript-unexpected-control-characters-in-regular-expression

It's needed because of the ansi escaping and you are right I must not have run format

**Comment by Keyrxng** [comment]
@gentlementlegen I've cleaned things up, bumped ts and the plugins being used and removed those that were not and updated the wf 

Seems that no dependencies are required according to Knip and from running the tests after everything seems to be working so I've removed those too.

Removed what comments were not needed, the code is pretty straight forward so not many are needed but I've left some

File: src/utils.ts (hunk 5 above)
Comment (gentlementlegen) [reply src/utils.ts:1]: I was referring to the eslint rule disable? Does this regex cause any issue?

Comment (Keyrxng) [reply src/utils.ts:1]: No issues, it does the job it's supposed to do 

> Control characters are special, invisible characters in the ASCII range 0-31. These characters are rarely used in JavaScript strings so a regular expression containing elements that explicitly match these characters is most likely a mistake.

From the eslint docs, but in this instance it's not a mistake

**Review by gentlementlegen — APPROVED** [review]

File: .github/workflows/release-please.yml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+permissions:
+  contents: write
+  pull-requests: write
+
+jobs:
+  release-please:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: google-github-actions/release-please-action@v3
```
Comment (gentlementlegen) [inline .github/workflows/release-please.yml]: This package has been deprecated, if you can please update it to `googleapis`

**Review by gentlementlegen — CHANGES_REQUESTED** [review]

File: .github/workflows/release-please.yml (hunk 6 above)
Comment (Keyrxng) [reply .github/workflows/release-please.yml]: https://github.com/googleapis/release-please-action?tab=readme-ov-file#basic-configuration

The readme includes a comment about setting up a PAT token as the secret

**Review by gentlementlegen — APPROVED** [review]
